sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java

```diff
@@ -0,0 +1,314 @@
+import org.opensearch.ResourceAlreadyExistsException;
+import org.opensearch.action.admin.cluster.health.ClusterHealthRequest;
+import org.opensearch.action.admin.cluster.health.ClusterHealthResponse;
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.health.ClusterHealthStatus;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.metadata.MappingMetadata;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.common.util.concurrent.ThreadContext.StoredContext;
+import org.opensearch.env.Environment;
+import org.opensearch.threadpool.ThreadPool;
+
+public class ConfigurationRepository {
+    private static final Logger LOGGER = LogManager.getLogger(ConfigurationRepository.class);
+
+    private final String securityIndex;
+    private final Client client;
+    private final Map<CType, SecurityDynamicConfiguration<?>> configCache = new HashMap<>();
+    private final List<ConfigurationChangeListener> configurationChangedListener;
+    private final ConfigurationLoader cl;
+    private final Settings settings;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private DynamicConfigFactory dynamicConfigFactory;
+    private static final int DEFAULT_CONFIG_VERSION = 1;
+    private final Thread bgThread;
+    private final AtomicBoolean installDefaultConfig = new AtomicBoolean();
+
+    private ConfigurationRepository(
+        Settings settings,
+        final Path configPath,
+        ThreadPool threadPool,
+        Client client,
+        ClusterService clusterService
+    ) {
+        this.securityIndex = settings.get(ConfigConstants.IDENTITY_CONFIG_INDEX_NAME, ConfigConstants.IDENTITY_DEFAULT_CONFIG_INDEX);
+        this.settings = settings;
+        this.client = client;
+        this.threadPool = threadPool;
+        this.clusterService = clusterService;
+        this.configurationChangedListener = new ArrayList<>();
+        cl = new ConfigurationLoader(client, threadPool, settings, clusterService);
+
+        bgThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    LOGGER.info("Background init thread started. Install default config?: " + installDefaultConfig.get());
+
+                    if (installDefaultConfig.get()) {
+
+                        try {
+                            String lookupDir = System.getProperty("identity.default_init.dir");
+                            final String cd = lookupDir != null
+                                ? (lookupDir + "/")
+                                : new Environment(settings, configPath).configDir().toAbsolutePath().toString() + "/";
+                            Path confFile = Path.of(cd + "internal_users.yml");
```

Comment:
this is a static list that will be loaded on server start, correct?
If so, does it mean when this config is uploaded the existing user indices, if any, will be deleted?
I saw the implementation of getConfigurationFromIndex but was still confused

Reply:
That's handled in `ConfigHelper.uploadFile`

```
try (Reader reader = createFileOrStringReader(cType, configVersion, filepath, populateEmptyIfFileMissing)) {

    final IndexRequest indexRequest = new IndexRequest(index).id(configType)
        .opType(OpType.CREATE)
        .setRefreshPolicy(RefreshPolicy.IMMEDIATE)
        .source(configType, readXContent(reader, XContentType.YAML));
    IndexResponse response = tc.index(indexRequest).actionGet();
    final String res = response.getId();

    if (!configType.equals(res)) {
        throw new Exception(
            "   FAIL: Configuration for '" + configType + "' failed for unknown reasons. Pls. consult logfile of opensearch"
        );
    }
    LOGGER.info("Doc with id '{}' and version {} is updated in {} index.", configType, configVersion, index);
} catch (VersionConflictEngineException versionConflictEngineException) {
    LOGGER.info("Index {} already contains doc with id {}, skipping update.", index, configType);
}
```

If it exists, then a `VersionConflictEngineException` is thrown and handled by skipping.